File: firstCodebase/old-code/detection-system-simulator.js
const fs = require('fs');
const path = require('path');
const { COLLECTIONS, addDocument, getLatestDocument, getDocumentsSorted } = require('./firebaseConfig');

// Path to sample image - we'll use a placeholder image for testing
const SAMPLE_IMAGE_PATH = path.join(__dirname, 'sample-shooter.png');

// Simulate coordinates within a reasonable range
function generateRandomCoordinates() {
  return {
    x: Math.floor(Math.random() * 1000),  // Random X between 0-1000
    y: Math.floor(Math.random() * 1000),  // Random Y between 0-1000
    timestamp: Date.now()
  };
}

// FUNCTION 1: Send a shooter image to the database
async function sendShooterImage() {
  try {
    // Check if the sample image exists
    if (!fs.existsSync(SAMPLE_IMAGE_PATH)) {
      console.error(`Sample image not found at: ${SAMPLE_IMAGE_PATH}`);
      console.log("Creating a valid placeholder image...");
      
      // Create a simple PNG image (1x1 pixel, red color)
      const pngData = Buffer.from([
        0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A, 0x00, 0x00, 0x00, 0x0D,
        0x49, 0x48, 0x44, 0x52, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x01,
        0x08, 0x02, 0x00, 0x00, 0x00, 0x90, 0x77, 0x53, 0xDE, 0x00, 0x00, 0x00,
        0x0C, 0x49, 0x44, 0x41, 0x54, 0x08, 0xD7, 0x63, 0xF8, 0xCF, 0xC0, 0x00,
        0x00, 0x03, 0x01, 0x01, 0x00, 0x18, 0xDD, 0x8D, 0xB0, 0x00, 0x00, 0x00,
        0x00, 0x49, 0x45, 0x4E, 0x44, 0xAE, 0x42, 0x60, 0x82
      ]);
      
      fs.writeFileSync(SAMPLE_IMAGE_PATH, pngData);
      console.log("Valid PNG image created");
    }
    
    // Read the sample image file
    const imageBuffer = fs.readFileSync(SAMPLE_IMAGE_PATH);
    const base64Data = imageBuffer.toString('base64');
    
    console.log(`[DEBUG] Image data size: ${base64Data.length} characters`);
    console.log(`[DEBUG] First 50 chars of base64: ${base64Data.substring(0, 50)}...`);
    
    // Save to database using Firebase directly
    const imageDoc = await addDocument(COLLECTIONS.SHOOTER_IMAGE, {
      base64Data: base64Data,
      timestamp: Date.now()
    });
    
    console.log(`Shooter image sent to database. ID: ${imageDoc.id}`);
    return imageDoc.id;
  } catch (error) {
    console.error("Error sending shooter image:", error);
    return null;
  }
}

// FUNCTION 2: Check for verification status
async function checkVerificationStatus() {
  try {
    // Get the latest verification document
    const latestVerification = await getLatestDocument(COLLECTIONS.SHOOTER_VERIFICATION);
    
    if (latestVerification && latestVerification.isDetected) {
      console.log("SHOOTER VERIFIED! isDetected = true");
      console.log("Proceeding with shooter classification...");
      return true;
    } else {
      const status = latestVerification ? 
        `rejected (isDetected = ${latestVerification.isDetected})` : 
        "not verified yet (no status)";
      console.log(`Shooter detection ${status}`);
      return false;
    }
  } catch (error) {
    console.error("Error checking verification status:", error);
    return false;
  }
}

// FUNCTION 3: Send shooter coordinates
async function sendShooterCoordinates() {
  try {
    // Generate random coordinates
    const coordinates = generateRandomCoordinates();
    
    // Save to database using Firebase directly
    await addDocument(COLLECTIONS.SHOOTER_COORDINATES, coordinates);
    
    console.log(`Coordinates sent: X=${coordinates.x}, Y=${coordinates.y}`);
    return coordinates;
  } catch (error) {
    console.error("Error sending coordinates:", error);
    return null;
  }
}

// Run the simulation
async function runSimulation() {
  console.log("Starting detection system simulation...");
  
  // First, send an initial image immediately
  console.log("\n--- IMMEDIATELY SENDING SHOOTER IMAGE ---");
  const imageId = await sendShooterImage();
  
  if (imageId) {
    console.log(`Image with ID ${imageId} sent to database`);
    console.log("The React Native app should now display this image");
  } else {
    console.error("Failed to send initial image");
  }
  
  // Send initial coordinates
  console.log("\n--- SENDING INITIAL COORDINATES ---");
  const initialCoords = await sendShooterCoordinates();
  if (initialCoords) {
    console.log(`Initial coordinates sent: X=${initialCoords.x}, Y=${initialCoords.y}`);
  } else {
    console.log('Failed to send initial coordinates');
  }
  
  // Check verification status initially
  console.log("\n--- CHECKING INITIAL VERIFICATION STATUS ---");
  await checkVerificationStatus();
  
  // Set up recurring tasks
  console.log("\n--- STARTING CONTINUOUS MONITORING ---");
  
  // Check verification status every 3 seconds
  const verificationInterval = setInterval(async () => {
    const isVerified = await checkVerificationStatus();
    
    // If verified, we can stop checking
    if (isVerified) {
      console.log("Shooter has been verified, stopping verification checks.");
      clearInterval(verificationInterval);
    }
  }, 3000);
  
  // Send coordinates every 2 seconds
  const coordinatesInterval = setInterval(() => {
    sendShooterCoordinates();
  }, 2000);
  
  // After 1 minute, stop the simulation
  setTimeout(() => {
    clearInterval(coordinatesInterval);
    console.log("\n--- SIMULATION COMPLETE ---");
    console.log("In a real scenario, the detection system would continue running.");
    console.log("Press Ctrl+C to exit.");
    // Firebase connections are managed automatically
  }, 60000);
}

// Start the simulation immediately
console.log("=".repeat(50));
console.log("SHOOTER DETECTION SYSTEM SIMULATOR");
console.log("This simulator immediately sends a shooter image and coordinates");
console.log("=".repeat(50));
runSimulation(); 

File: firstCodebase/detection_system_simulator.py
"""
detection_system_simulator.py
Python version of the shooter detection system simulator
"""

import os
import time
import random
import base64
from PIL import Image
import io
import threading
from firebase_config import COLLECTIONS, add_document, get_latest_document

# Path to sample image - we'll use a placeholder image for testing
SAMPLE_IMAGE_PATH = os.path.join(os.path.dirname(__file__), 'sample-shooter.png')

def generate_random_coordinates():
    """Generate random coordinates within a reasonable range"""
    return {
        'x': random.randint(0, 1000),  # Random X between 0-1000
        'y': random.randint(0, 1000),  # Random Y between 0-1000
        'timestamp': int(time.time() * 1000)  # Current time in milliseconds
    }

def create_placeholder_image():
    """Create a simple red square image if the sample image doesn't exist"""
    print("Creating a valid placeholder image...")
    
    # Create a 100x100 red square image
    img = Image.new('RGB', (100, 100), color='red')
    
    # Save the image
    img.save(SAMPLE_IMAGE_PATH)
    print("Valid PNG image created")

def send_shooter_image():
    """Send a shooter image to the database"""
    try:
        # Check if the sample image exists
        if not os.path.exists(SAMPLE_IMAGE_PATH):
            create_placeholder_image()
        
        # Read the sample image file
        with open(SAMPLE_IMAGE_PATH, 'rb') as image_file:
            image_bytes = image_file.read()
            base64_data = base64.b64encode(image_bytes).decode('utf-8')
        
        print(f"[DEBUG] Image data size: {len(base64_data)} characters")
        print(f"[DEBUG] First 50 chars of base64: {base64_data[:50]}...")
        
        # Create a data URI format that the app expects
        # The app is expecting a data URI format like 'data:image/png;base64,...'
        # but we're just sending the raw base64 data
        
        # Save to database using Firebase
        image_doc = add_document(COLLECTIONS.SHOOTER_IMAGE, {
            'base64Data': base64_data,
            'timestamp': int(time.time() * 1000)
        })
        
        print(f"Shooter image sent to database. ID: {image_doc['id']}")
        return image_doc['id']
    
    except Exception as error:
        print(f"Error sending shooter image: {error}")
        return None

def check_verification_status():
    """Check for verification status"""
    try:
        # Get the latest verification document
        latest_verification = get_latest_document(COLLECTIONS.SHOOTER_VERIFICATION)
        
        if latest_verification and latest_verification.get('isDetected'):
            print("SHOOTER VERIFIED! isDetected = true")
            print("Proceeding with shooter classification...")
            return True
        else:
            status = f"rejected (isDetected = {latest_verification.get('isDetected')})" if latest_verification else "not verified yet (no status)"
            print(f"Shooter detection {status}")
            return False
    
    except Exception as error:
        print(f"Error checking verification status: {error}")
        return False

def send_shooter_coordinates():
    """Send shooter coordinates"""
    try:
        # Generate random coordinates
        coordinates = generate_random_coordinates()
        
        # Save to database using Firebase
        add_document(COLLECTIONS.SHOOTER_COORDINATES, coordinates)
        
        print(f"Coordinates sent: X={coordinates['x']}, Y={coordinates['y']}")
        return coordinates
    
    except Exception as error:
        print(f"Error sending coordinates: {error}")
        return None

def run_simulation():
    """Run the detection system simulation"""
    print("Starting detection system simulation...")
    
    # First, send an initial image immediately
    print("\n--- IMMEDIATELY SENDING SHOOTER IMAGE ---")
    image_id = send_shooter_image()
    
    if image_id:
        print(f"Image with ID {image_id} sent to database")
        print("The React Native app should now display this image")
    else:
        print("Failed to send initial image")
    
    # Send initial coordinates
    print("\n--- SENDING INITIAL COORDINATES ---")
    initial_coords = send_shooter_coordinates()
    if initial_coords:
        print(f"Initial coordinates sent: X={initial_coords['x']}, Y={initial_coords['y']}")
    else:
        print('Failed to send initial coordinates')
    
    # Check verification status initially
    print("\n--- CHECKING INITIAL VERIFICATION STATUS ---")
    check_verification_status()
    
    # Set up recurring tasks
    print("\n--- STARTING CONTINUOUS MONITORING ---")
    
    # Flag to control the threads
    running = True
    
    # Function to check verification status periodically
    def verification_checker():
        while running:
            is_verified = check_verification_status()
            
            # If verified, we can stop checking
            if is_verified:
                print("Shooter has been verified, stopping verification checks.")
                break
            
            time.sleep(3)  # Check every 3 seconds
    
    # Function to send coordinates periodically
    def coordinates_sender():
        while running:
            send_shooter_coordinates()
            time.sleep(2)  # Send every 2 seconds
    
    # Start the threads
    verification_thread = threading.Thread(target=verification_checker)
    coordinates_thread = threading.Thread(target=coordinates_sender)
    
    verification_thread.daemon = True
    coordinates_thread.daemon = True
    
    verification_thread.start()
    coordinates_thread.start()
    
    try:
        # Run for 60 seconds or until Ctrl+C
        time.sleep(60)
        print("\n--- SIMULATION COMPLETE ---")
        print("In a real scenario, the detection system would continue running.")
        print("Press Ctrl+C to exit.")
        
        # Keep the main thread running to allow for Ctrl+C
        while True:
            time.sleep(1)
    
    except KeyboardInterrupt:
        # Set the flag to stop the threads
        running = False
        print("\nSimulation stopped by user.")

if __name__ == "__main__":
    # Print banner
    print("=" * 50)
    print("SHOOTER DETECTION SYSTEM SIMULATOR (PYTHON)")
    print("This simulator immediately sends a shooter image and coordinates")
    print("=" * 50)
    
    # Start the simulation
    run_simulation()
